File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char,char> m;
        m[')'] = '(';
        m[']'] = '[';
        m['}'] = '{';
        for(auto ele: s){
            if(ele == '(' || ele == '{' || ele == '[') st.push(ele);
            else {
                if(st.empty() || m[ele] != st.top()) return false;
                st.pop();
            }
        }
        if(!st.empty()) return false;
        return true;
    }
};

File: 20-valid-parentheses/valid-parentheses.py
from collections import deque

class Solution:
    def isValid(self, s: str) -> bool:
        q = deque()
        for ch in s:
            if ch in ('(','[','{'):
                q.append(ch)
            else:
                if (len(q) > 0) and ((ch == ')' and q[-1] == '(') or (ch == ']' and q[-1] == '[') or (ch == '}' and q[-1] == '{')):
                    q.pop()
                else:
                    return False
        
        if len(q) > 0:
            return False
        return True
                